client/player/wasapi_player.cpp

```diff
@@ -0,0 +1,469 @@
+	waveformatExtended->Format                      = *waveformat;
+	waveformatExtended->Format.wFormatTag           = WAVE_FORMAT_EXTENSIBLE;
+	waveformatExtended->Format.cbSize               = 22;
+	waveformatExtended->Samples.wValidBitsPerSample = waveformat->wBitsPerSample;
+	waveformatExtended->dwChannelMask               = SPEAKER_FRONT_LEFT | SPEAKER_FRONT_RIGHT;
+	waveformatExtended->SubFormat                   = KSDATAFORMAT_SUBTYPE_PCM;
+
+
+	// Retrieve the device enumerator
+	IMMDeviceEnumeratorPtr deviceEnumerator = nullptr;
+	hr = CoCreateInstance(
+		CLSID_MMDeviceEnumerator, NULL, 
+		CLSCTX_SERVER, IID_IMMDeviceEnumerator,
+		(void**)&deviceEnumerator);
+	CHECK_HR(hr);
+
+	// Register the default playback device (eRender for playback)
+	IMMDevicePtr device = nullptr;
+	if (pcmDevice_.idx == 0)
+	{
+		hr = deviceEnumerator->GetDefaultAudioEndpoint(eRender, eConsole, &device);
+		CHECK_HR(hr);
+	}
+	else
+	{
+		IMMDeviceCollectionPtr devices = nullptr;
+		hr = deviceEnumerator->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, &devices);
+		CHECK_HR(hr);
+
+		devices->Item(pcmDevice_.idx, &device);
+	}
+
+	IPropertyStorePtr properties = nullptr;
+	hr = device->OpenPropertyStore(STGM_READ, &properties);
+	CHECK_HR(hr);
+
+	PROPVARIANT format;
+	hr = properties->GetValue(PKEY_AudioEngine_DeviceFormat, &format);
+	CHECK_HR(hr);
+
+	PWAVEFORMATEX formatEx = (PWAVEFORMATEX)format.blob.pBlobData;
+	LOG(INFO, LOG_TAG) << "Device accepts format: " << formatEx->nSamplesPerSec << ":" << formatEx->wBitsPerSample << ":" << formatEx->nChannels << "\n";
+
+	// Activate the device
+	IAudioClientPtr audioClient = nullptr;
+	hr = device->Activate(IID_IAudioClient, CLSCTX_SERVER, NULL, (void**)&audioClient);
+	CHECK_HR(hr);
+
+	hr = audioClient->IsFormatSupported(
+		AUDCLNT_SHAREMODE_EXCLUSIVE,
+		&(waveformatExtended->Format),
+		NULL);
+	CHECK_HR(hr);
+
+	IAudioSessionManagerPtr sessionManager = nullptr;
+    // Get the session manager for the endpoint device.
+    hr = device->Activate(__uuidof(IAudioSessionManager), CLSCTX_INPROC_SERVER, NULL, (void**)&sessionManager);
+    CHECK_HR(hr);
+
+    // Get the control interface for the process-specific audio
+    // session with session GUID = GUID_NULL. This is the session
+    // that an audio stream for a DirectSound, DirectShow, waveOut,
+    // or PlaySound application stream belongs to by default.
+    IAudioSessionControlPtr control = nullptr;
+    hr = sessionManager->GetAudioSessionControl(NULL, 0, &control);
+    CHECK_HR(hr);
+
+    // register
+    hr = control->RegisterAudioSessionNotification(audioEventListener_);
+    CHECK_HR(hr);
+	
+	// Get the device period
+	REFERENCE_TIME hnsRequestedDuration = REFTIMES_PER_SEC;
+	hr = audioClient->GetDevicePeriod(NULL, &hnsRequestedDuration);
+	CHECK_HR(hr);
+	
+	// Initialize the client at minimum latency
+	hr = audioClient->Initialize(
+		AUDCLNT_SHAREMODE_EXCLUSIVE,
```

Comment:
sorry - I mentioned in the issue thread that the shared mode feature was in a separate branch in case you wanted to review it separately, then forgot about it :) since shared mode will be the default for windows anyway I'll just merge that in master now